### DISCUSSION TITLE
Keep SSH sessions alive.

### DISCUSSION THREAD

**Discussion by axkibe** [libgit2/libgit2#6177] [General] [answered]
I'm now for a few days investigating how one can keep SSH session alive if communicating with the same server for multiple repositories (and open new SSH channels while keeping the authentication of the one master session alive, this usually _tremendously_ improves speed as ssh authentication negotiation is the bottleneck). In plain commandline git/ssh mode something like this is achieved with what ssh calls "ControlMaster" in .ssh/config.

I kind of achieved this by hacking gitlib2 and adding a new "ssh_session_hook".
https://github.com/axkibe/libgit2/blob/sessionhook/src/transports/ssh.c#L559
Allowing the application to set the ssh session to an existing one. And also a hook in ssh_stream_free
https://github.com/axkibe/libgit2/blob/sessionhook/src/transports/ssh.c#L215
To stop gitlib2 to close the first session by setting the pointers NULL.

However IMO this is a dirty hack, thus I didn't even submit a PR.

I'd much rather just be able to overload _git_ssh_setup_conn() and ssh_stream_free() in the main application for a version that tracks ssh hosts and keeps sessions alive if multiple requests are to come. (that is a mass pull from a lot of repositories).

Unfortunally the API doesn't seem to allow that. Of course one could provide a complete own transport, but that would duplicate a lot of code and be much to much overkill for a rather small adaption.

There is the option to provide a transport callback as fetch paramter and request a git_transport_new() and then modify the function pointers.

```int transport_cb(git_transport **out, git_remote *owner, void *param)
{
    git_transport *t;
    git_transport_new( &t, owner, "ssh://localhost//home/axel/gits/test1" );
    t->wrapped->action = my_ssh_action;
    *out = t;
    return 0;
}
```

Unfortunately, and here it gets ugly, `wrapped` as sub_transport isn't visible in git_transport, it's part of transport_smart which would need a typecast but transport_smart isn't exported in include/git2 at all. What in my_ssh_action I would call my_ssh_uploadpack_ls() which would use "my_connect" which contrary to the gitlib2 implemenation would save established SSH sessions for timely reuse (my case all channels would happen with only one host anyway).

PS: to complicate things, my driver was to written in rust anyway, so I'd need to adapt that API also.

Any ideas how to be approach this in a cleaner manner?

Other than providing a new callback, or allow to overload function pointers to make the whole system more LEGO/toolbox like, there would the option to build this functionality directly into the lib. like "keep_alive = true" but then would need to manage a data structure of open SSH sessions.

To be honest I'm on the verge of giving up on this and for now spawning the git command line binary instead. With ssh control masters that would likely be faster than a native lib implementation that reconnects for every repository. 

TLDR: What I'm kind of getting at, what I'm missing is a layering of having the host application taking total care of establishing and closing the SSH connection itself, but having libgit2 control over what is happening in the channel. e.g. also  #4872 we be solved by this.

**Comment by boretrk** [comment]
I don't know anything about how the transport stuff is done but #4872 references PR #4881.
It is a bit outdated but maybe there is something there that can help you out?

**Comment by ethomson** [comment, marked as answer]
I'm not sure that I have great answers for you here.  I was just in the remote and transport code, and even doing something seemingly simple was rather maddening (https://github.com/libgit2/libgit2/pull/6175).  It feels like it hasn't aged particularly well and could use a serious refresher.  I have some ideas of improvements to make but it's also not at the top of my priority list at the moment.

But one of the things that nags at me is the way that the smart transports are layered.  I have considered trying to invert the logic so that http and ssh are transports and call into a smart "helper".  Or maybe just do some refactoring on the "subclass" notion that's there.  What's there I find pretty tricky to reason about, there's a lot of indirection going on.

If you had a notion to improve this, I'd certainly be interested in looking at a plan with you, or reviewing your code.  I think that some outside eyes would be helpful.

In particular, I think that my rather naive thought here is that allowing you to take an ssh transport and modify it or further subclass it to achieve your goals feels like the "right" solution here.  It's not that ControlMaster or ProxyCommand are particularly odd...  but it does feel like exposing a lot of ssh knobs is going to get real complex real fast.  We already have a lot and it already sort of ties us to libssh2 as our transport and makes adding something like libssh as a second alternative challenging.  (We have this problem with TLS systems as well, but in a sense TLS is rather easier since there are fewer options to expose here.)

Having said that...  if I were in your shoes, I would _also_ start wondering if libssh2 is really the right solution here since this is already baked in to ssh itself.  I have very little experience with libssh2 to really have a good idea of its quality.  (I've used libssh in production and have a reasonable amount of trust in it.)  And if I knew that ssh was available on my system and I was executing a command-line application that could prompt people when `WARNING HOST KEY HAS CHANGED`, well, then, just running `/usr/bin/ssh` feels like a reasonable idea to me.

In fact, I noodled around on a smart subtransport that called ssh: https://github.com/libgit2/libgit2/pull/5507/commits/de499cc21335ac9ffe44e6c1825ff7417b45a895

I think that it's worth including in libgit2 in the long term but the "where" and "how" needs a little more thought.

(Also, a minor nit, the project is _libgit2_ not _gitlib2_.)

**Comment by axkibe** [comment]
Thank you, ethomson, now having spent a few days with the code this was my impression myself. It really feels nice to hear a similar opinion on this. Things with transports just feel weird. I know this from my own projects when designing some code, keeping modifying it, and when I feel for another project, "i already done this i can easily adapt that old code", working through it for slightly different aims makes me feel, it just became weird when continuously fixing small stuff.

If I'm frank, I'm not too found about the callbacks as basic principle, same with other projects that keep putting more and more hooks in all places, that's also why I don't want to push the idea of a (yet another) session hook. I basically like more "LEGO" style systems, where as user you can build your own thing given easy building blocks  like (here make a connection, here establish a git request etc.) in contrast to having the "this is the one solution, but there are a million nobs you can turn". But I'm not in the position to even approach a major refractory. Especially given this project it's success as many applications it has.

You're completely right, now after a few days trying to get this to work, I decided the same, I'll just spawn /usr/bin/git (and use GIT_SSH_COMMAND environment to hand it a control master I spawned before), I'll keep using libgit2* for local queries (like probing what branch a work tree is at, if it's clean etc.) but everything network related, I'll hand of to the command line tool. 

* (and sorry for the missspell, I also know how this feel when people mix up my project names ;)
PS: Sorry don't count me in for some serious work, I was just in quickly for this one task in mind and I kinda failed ;)

**Comment by musabshak** [comment]
@axkibe Is the conclusion to this discussion that using `libgit2`, you can't do SSH "connection pooling" like you can with `/usr/bin/ssh` commands by setting:
```
Host *
    ControlMaster auto
    ControlPath ~/.ssh/sockets/%r@%h-%p
    ControlPersist 10m
```
in `~/.ssh/config`? That is any `git fetch` etc equivalent commands you do using `libgit2` will open a brand new SSH connection and kill it immediately afterwards? Thanks!

(just trying to develop an understanding of `libgit2`)
